### PR TITLE
New Module: vmware_register - Register/un-register VMs and templates

### PR DESCRIPTION
**Pull request by garbled1**
##### SUMMARY

Add a new module to allow registering of files in vmware as guests or templates.  Also allows you to un-register the guest/template.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/vmware/vmware_register.py

##### ANSIBLE VERSION

```
ansible 2.5.0 (vmware_register 3a8c949ddd) last updated 2017/09/30 21:14:07 (GMT +000)
  config file = None
  configured module search path = [u'/home/ec2-user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ec2-user/ansible/lib/ansible
  executable location = /home/ec2-user/ansible/bin/ansible
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### ADDITIONAL INFORMATION

```
- name: Register a template
  vsphere_register:
    state: present
    name: My_Template
    is_template: True
    path: templates/template.vmxt
    folder: My_Templates/Template
    datacenter: dc_1
    datastore: ds_1
    hostname: vsphere.host.com
    username: administrator@vsphere.local
    password: vmware
    validate_certs: false
  delegate_to: localhost

- name: Delete VM from inventory
  vsphere_register:
    state: absent
    name: vm_guest_name
    path: vm/vm_guest_name/vm_guest_name.vmdk
    folder: Databases/Production
    datacenter: dc_1
    datastore: prod_ds
    hostname: vsphere.host.com
    username: administrator@vsphere.local
    password: vmware
    validate_certs: false
  delegate_to: localhost

- name: Register a VM to a specific esxi host
  vsphere_register:
    state: present
    name: new_db_host
    is_template: True
    path: vm/new_db_host/new_db_host.vmdk
    folder: Databases/Production
    datacenter: dc_1
    datastore: ds_1
    esxi_hostname: prod_db_esxi.my.com
    hostname: vsphere.host.com
    username: administrator@vsphere.local
    password: vmware
    validate_certs: false
  delegate_to: localhost
```

### CODE REVIEW

**Comment by Akasurde** [comment]
@garbled1 Could you please integration test for this change ?

**Comment by garbled1** [comment]
I'm more than willing to.  I'm staring at the tests for vmware_guest now, and it's not obvious to me what the structure of vcsim is.   For example, how would I know what the datacenter/folder and path arguments are to find the VM I want to register/unregister.  Where is the template?  Is the vcsim readily available, ie, could I poke at it from home here and find out?

**Comment by Akasurde** [comment]
@garbled1 You can find details about vcsim infrastructure deployment [here](https://github.com/ansible/community/wiki/VMware%3A-govcsim).

**Comment by garbled1** [comment]
Tests added.  :)

**Comment by garbled1** [comment]
I'm not sure I understand this failure at all.

```
  File "/tmp/ansible_tv7vnod0/ansible_module_vmware_register.py", line 257, in obj_has_parent
    current_parent = current_parent.parent
  File "/usr/local/lib/python3.5/dist-packages/pyVmomi/VmomiSupport.py", line 574, in __call__
    return self.f(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/pyVmomi/VmomiSupport.py", line 394, in _InvokeAccessor
    return self._stub.InvokeAccessor(self, info)
  File "/usr/local/lib/python3.5/dist-packages/pyVmomi/StubAdapterAccessorImpl.py", line 43, in InvokeAccessor
    objectContent = result.objects[0]
IndexError: list index out of range
```
Is something different happening in python 3.5 wrt that parent object?

File: lib/ansible/modules/cloud/vmware/vmware_register.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+                break
+            obj = obj.parent
+            if isinstance(obj, vim.Datacenter):
+                datacenter = obj
+                break
+        self.parent_datacenters[obj] = datacenter
+        return datacenter
+
+
+class PyVmomiHelper(object):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_register.py:195]: can you implement [PyVmomi](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/vmware.py#L745) ?

File: lib/ansible/modules/cloud/vmware/vmware_register.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+            except vim.fault.InvalidPowerState:
+                module.fail_json(msg="Cannot unregister a VM which is powered on")
+            except Exception as e:
+                module.fail_json(msg=to_native(e))
+            result['changed'] = True
+            module.exit_json(**result)
+        else:
+            module.exit_json(**result)
+    else:
+        result = pyv.register_vm(template=module.params['is_template'])
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_register.py:408]: What happens if there is no VM available with user provided name ?

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@garbled1 Any news?

**Comment by garbled1** [comment]
Yeah, my testing infrastructure is moving about right now, I'm working on your suggestions but need to test it against real HW first.Hopefully in the next 2 weeks.

**Comment by Akasurde** [comment]
@garbled1 Are you still working on this ? If not let me know I will take over this PR.

**Comment by Akasurde** [comment]
@garbled1 Any news ?

**Comment by Akasurde** [comment]
@garbled1 Are you still working on this, else let me know I will take over this PR?

**Comment by Akasurde** [comment]
@garbled1 Any news ?

**Comment by Akasurde** [comment]
@garbled1 Are you still working on this ?

**Comment by Akasurde** [comment]
Closing this issue due to inactivity from contributor.

**Pull request closed without merging**